File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
class Solution:
    # frequency counter map: space: O(n), time: O(nlogn) to sort the map values and return k top or bottom
    # use priority queue. heap size always k. at the end return all values of heap to res. 

    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = collections.Counter(nums)
        
        heap = []
        
        for num in count.keys():
            heapq.heappush(heap, (count[num], num))
            if len(heap) > k:
                heapq.heappop(heap)
            
        res = []
        for i in range(len(heap)):
            res.append(heapq.heappop(heap)[1])
        return res
        

File: Data Structures & Algorithms/anagram-groups/submission-2.go
func groupAnagrams(strs []string) [][]string {

	res := make(map[[26]int][]string)

	for _, s:= range strs {
		var count [26]int
		for _, c:= range s {
			count[c-'a']++
		}
		res[count] = append(res[count], s)
	}

	var result [][]string
	for _, val := range res {
		result = append(result, val)
	}
	return result
	
}
